File: research/solver.py
from itertools import combinations

Point = tuple[int, int]
PenSet = list[tuple[int, int]]

_COLORS = [31, 32, 33, 34, 35, 36, 91, 92, 93, 94, 95, 96, 97, 90]


def _colored(text: str, color: int) -> str:
    return f"\033[{color}m{text}\033[0m"


class Solver:
    def __init__(self, board: list[list[int]]):
        self.board = board
        self.SIZE = len(board)
        self.mask: set[Point] = set()
        self.pensets: list[PenSet] = self._get_pen_sets()
        self._solution: list[Point] = []
        self.had_to_guess = False

    def solve(self) -> list[Point]:
        """Returns one bull position per region, or [] if unsolvable."""
        if self._solve():
            self._solution = [ps[0] for ps in self.pensets if ps]
            return self._solution
        return []

    # --- display ---

    def show_board(self):
        """Prints the board with each region in its own color."""
        print()
        for y in range(self.SIZE):
            for x in range(self.SIZE):
                region = self.board[y][x]
                print(_colored(str(region), _COLORS[region % len(_COLORS)]), end=" ")
            print()

    def show_mask(self):
        """Prints the board with region colors and masked cells shown as #."""
        print()
        for y in range(self.SIZE):
            for x in range(self.SIZE):
                if (y, x) in self.mask:
                    print(_colored("#", 90), end=" ")
                else:
                    region = self.board[y][x]
                    print(
                        _colored(str(region), _COLORS[region % len(_COLORS)]), end=" "
                    )
            print()

    def show_solution(self):
        """Prints the board with bull positions marked as ★."""
        solution_set = set(self._solution)
        print()
        for y in range(self.SIZE):
            for x in range(self.SIZE):
                region = self.board[y][x]
                color = _COLORS[region % len(_COLORS)]
                mark = "★" if (y, x) in solution_set else str(region)
                print(_colored(mark, color), end=" ")
            print()

    # --- backtracking ---

    def _solve(self) -> bool:
        prev_size = -1
        while len(self.mask) != prev_size:
            prev_size = len(self.mask)
            self._single_pen()
            self._one_direction()
            self._pen_overlap()
            self._overcounting()
            if any(len(ps) == 0 for ps in self.pensets):
                return False
            if len(self.mask) == self.SIZE**2 - self.SIZE:
                return True

        # Stuck: pick region with fewest candidates (must have ≥2 to make a real choice)
        self.had_to_guess = True
        target = min((ps for ps in self.pensets if len(ps) >= 2), key=len)

        for y, x in list(target):
            saved_mask = set(self.mask)
            saved_pensets = [list(ps) for ps in self.pensets]

            for cy, cx in list(target):
                if (cy, cx) != (y, x):
                    self.mask.add((cy, cx))
            for pt in self._get_adjacent(y, x):
                self.mask.add(pt)
            self._readjust_pensets()

            if self._solve():
                return True

            self.mask.clear()
            self.mask.update(saved_mask)
            for i, ps in enumerate(self.pensets):
                ps.clear()
                ps.extend(saved_pensets[i])

        return False

    # --- strategies ---

    def _single_pen(self):
        for penset in self.pensets:
            if len(penset) == 1:
                for pt in self._get_adjacent(*penset[0]):
                    self.mask.add(pt)
        self._readjust_pensets()

    def _one_direction(self):
        for penset in self.pensets:
            if not penset:
                continue
            vert = self._penset_all_vert(penset)
            if vert is not None:
                x, color = vert
                for y in range(self.SIZE):
                    if self.board[y][x] != color:
                        self.mask.add((y, x))
            horiz = self._penset_all_horiz(penset)
            if horiz is not None:
                y, color = horiz
                for x in range(self.SIZE):
                    if self.board[y][x] != color:
                        self.mask.add((y, x))
        self._readjust_pensets()

    def _pen_overlap(self):
        for penset in self.pensets:
            if not penset or len(penset) > 8:
                continue
            intersect = set(self._get_adjacent(*penset[0]))
            for y, x in penset:
                intersect &= set(self._get_adjacent(y, x))
            self.mask |= intersect
        self._readjust_pensets()

    def _overcounting(self):
        SIZE = self.SIZE
        row_regions: list[set] = [set() for _ in range(SIZE)]
        col_regions: list[set] = [set() for _ in range(SIZE)]
        color_to_ps: dict[int, PenSet] = {}

        for ps in self.pensets:
            if not ps:
                continue
            color = self.board[ps[0][0]][ps[0][1]]
            color_to_ps[color] = ps
            for y, x in ps:
                row_regions[y].add(color)
                col_regions[x].add(color)

        for k in range(1, SIZE):
            for row_subset in combinations(range(SIZE), k):
                regions: set = set()
                for y in row_subset:
                    regions |= row_regions[y]
                    if len(regions) > k:
                        break
                if len(regions) != k:
                    continue
                for color in regions:
                    for y, x in color_to_ps[color]:
                        if y not in row_subset:
                            self.mask.add((y, x))

            for col_subset in combinations(range(SIZE), k):
                regions = set()
                for x in col_subset:
                    regions |= col_regions[x]
                    if len(regions) > k:
                        break
                if len(regions) != k:
                    continue
                for color in regions:
                    for y, x in color_to_ps[color]:
                        if x not in col_subset:
                            self.mask.add((y, x))

        self._readjust_pensets()

    # --- utils ---

    def _readjust_pensets(self):
        for penset in self.pensets:
            for y, x in penset:
                if (y, x) in self.mask:
                    penset.remove((y, x))

    def _get_pen_sets(self) -> list[PenSet]:
        pen_sets: list[PenSet] = [[] for _ in range(self.SIZE)]
        for row in range(self.SIZE):
            for col in range(self.SIZE):
                pen_sets[self.board[row][col]].append((row, col))
        return pen_sets

    def _get_adjacent(self, y: int, x: int) -> list[Point]:
        out = []
        for dy in range(-1, 2):
            for dx in range(-1, 2):
                if dy == 0 and dx == 0:
                    continue
                ny, nx = y + dy, x + dx
                if 0 <= ny < self.SIZE and 0 <= nx < self.SIZE:
                    out.append((ny, nx))
        return out

    def _penset_all_horiz(self, penset: PenSet):
        ypos, xpos = penset[0]
        color = self.board[ypos][xpos]
        for i in range(1, len(penset)):
            if penset[i][0] != ypos:
                return None
        return ypos, color

    def _penset_all_vert(self, penset: PenSet):
        ypos, xpos = penset[0]
        color = self.board[ypos][xpos]
        for i in range(1, len(penset)):
            if penset[i][1] != xpos:
                return None
        return xpos, color


if __name__ == "__main__":
    # board = [
    #     [3, 3, 5, 5, 5, 4],
    #     [3, 3, 1, 5, 5, 0],
    #     [3, 3, 1, 1, 0, 0],
    #     [3, 3, 1, 0, 0, 0],
    #     [1, 1, 1, 0, 0, 0],
    #     [2, 0, 0, 0, 0, 0],
    # ]

    # board = [
    #     [0, 1, 1, 2, 2, 2],
    #     [0, 3, 2, 2, 2, 2],
    #     [0, 3, 3, 2, 2, 2],
    #     [3, 3, 4, 2, 2, 2],
    #     [4, 4, 4, 5, 5, 5],
    #     [4, 4, 4, 4, 5, 5],
    # ]

    # board = [
    #     [0, 0, 1, 1, 1, 1, 2, 2, 2, 3],
    #     [0, 1, 1, 1, 1, 1, 2, 2, 3, 3],
    #     [0, 1, 0, 1, 1, 1, 1, 3, 3, 4],
    #     [0, 0, 0, 1, 1, 5, 5, 5, 4, 4],
    #     [6, 6, 0, 0, 5, 5, 5, 5, 4, 4],
    #     [6, 6, 0, 0, 5, 7, 7, 7, 7, 8],
    #     [6, 6, 6, 6, 5, 7, 7, 7, 8, 8],
    #     [6, 6, 6, 9, 9, 9, 7, 7, 8, 8],
    #     [6, 6, 6, 6, 9, 9, 9, 9, 8, 8],
    #     [6, 6, 6, 6, 9, 9, 9, 9, 9, 8],
    # ]

    # board = [
    #     [0, 0, 1, 1],
    #     [0, 0, 1, 2],
    #     [3, 3, 1, 2],
    #     [3, 3, 2, 2],
    # ]

    board = [
        [0, 0, 0, 1, 1, 1, 2, 2],
        [0, 0, 0, 1, 1, 1, 1, 2],
        [3, 0, 0, 0, 1, 2, 1, 2],
        [3, 3, 0, 0, 2, 2, 2, 2],
        [3, 3, 3, 0, 2, 2, 2, 4],
        [3, 5, 5, 6, 6, 2, 4, 4],
        [7, 7, 5, 6, 6, 4, 4, 4],
        [5, 5, 5, 5, 4, 4, 4, 4],
    ]

    solver = Solver(board)
    solver.show_board()
    solver.solve()
    solver.show_board()
    solver.show_mask()
    print("Had to guess?", solver.had_to_guess)

File: src/solver.ts
import type { BullPen } from "./bullpen";

/** Point is [y, x] */
export type Point = [number, number];
type Region = Point[];

function hashPt(y: number, x: number): string {
    return `${y},${x}`;
}

function combinations(arr: number[], k: number): number[][] {
    if (k == 0) return [[]];
    if (arr.length < k) return [];
    const [first, ...rest] = arr;
    return [
        ...combinations(rest, k - 1).map(c => [first, ...c]),
        ...combinations(rest, k),
    ];
}

export class Solver {
    board: number[][];
    size: number;
    mask: Set<string>;
    regions: Region[];
    hadToGuess = false;

    private solution: Point[] = [];
    hintFns = [this.singlePen, this.oneDirection, this.penOverlap, this.overcounting];
    hintDesc = [
        "There is only one available spot in this pen",
        "All other cells would make it so there cannot be any bulls here",
        "Placing a bull here would make it so there cannot be any bulls here",
        "These regions encompass the entirety of these rows/cols"
    ];

    constructor(pen: BullPen) {
        this.board = pen.board;
        this.size = pen.size;
        this.mask = new Set();
        this.regions = this.getRegions();
    }

    solve(): Point[] {
        if (this._solve()) {
            this.solution = this.regions.filter(ps => ps.length > 0).map(ps => ps[0]);
            return this.solution;
        }
        return [];
    }

    nextHint(): { hint: string, dots: Point[] } | null {
        // TODO: check if the user made any mistakes

        let prevSize = this.mask.size;
        const oldMask = new Set(this.mask);

        for (let c = 0; c < this.hintFns.length; c++) {
            this.hintFns[c].bind(this)();
            if (prevSize != this.mask.size) {
                // this hint did something!
                const newPts: Point[] = [...this.mask.difference(oldMask)]
                    .map(k => k.split(',').map(Number) as Point);
                return { hint: this.hintDesc[c], dots: newPts };
            }
        }

        // just solve it, and then see...
        return null;
    }

    private _solve(): boolean {
        let prevSize = -1;
        while (this.mask.size != prevSize) {
            prevSize = this.mask.size;
            this.singlePen();
            this.oneDirection();
            this.penOverlap();
            this.overcounting();
            if (this.regions.some(ps => ps.length == 0)) return false;
            if (this.mask.size == this.size ** 2 - this.size) return true;
        }

        this.hadToGuess = true;
        const target = this.regions
            .filter(ps => ps.length >= 2)
            .reduce((a, b) => (a.length <= b.length ? a : b));

        for (const [y, x] of [...target]) {
            const savedMask = new Set(this.mask);
            const savedPensets = this.regions.map(ps => [...ps]);

            for (const [cy, cx] of [...target]) {
                if (cy != y || cx != x) this.mask.add(hashPt(cy, cx));
            }
            for (const [ay, ax] of this.getAdjacent(y, x)) {
                this.mask.add(hashPt(ay, ax));
            }
            this.readjustRegions();

            if (this._solve()) return true;

            this.mask.clear();
            for (const k of savedMask) this.mask.add(k);
            for (let i = 0; i < this.regions.length; i++) {
                this.regions[i].length = 0;
                this.regions[i].push(...savedPensets[i]);
            }
        }

        return false;
    }

    private singlePen() {
        for (const penset of this.regions) {
            if (penset.length == 1) {
                for (const [ay, ax] of this.getAdjacent(penset[0][0], penset[0][1])) {
                    this.mask.add(hashPt(ay, ax));
                }
            }
        }
        this.readjustRegions();
    }

    private oneDirection() {
        for (const penset of this.regions) {
            if (!penset.length) continue;
            const vert = this.pensetAllVert(penset);
            let changed = false;
            if (vert != null) {
                const [x, color] = vert;
                for (let y = 0; y < this.size; y++) {
                    if (this.board[y][x] != color) {
                        this.mask.add(hashPt(y, x));
                        changed = true;
                    }
                }
            }

            if (changed) {
                break;
            }

            const horiz = this.pensetAllHoriz(penset);
            if (horiz != null) {
                const [y, color] = horiz;
                for (let x = 0; x < this.size; x++) {
                    if (this.board[y][x] != color) {
                        this.mask.add(hashPt(y, x));
                    }
                }
            }
        }
        this.readjustRegions();
    }

    private penOverlap() {
        for (const penset of this.regions) {
            if (!penset.length || penset.length > 8) continue;
            let intersect = new Set(
                this.getAdjacent(penset[0][0], penset[0][1]).map(([ay, ax]) => hashPt(ay, ax))
            );
            for (const [y, x] of penset) {
                const adj = new Set(this.getAdjacent(y, x).map(([ay, ax]) => hashPt(ay, ax)));
                for (const k of [...intersect]) {
                    if (!adj.has(k)) intersect.delete(k);
                }
            }

            for (const k of intersect) this.mask.add(k);
        }
        this.readjustRegions();
    }

    private overcounting() {
        const SIZE = this.size;
        const rowRegions: Set<number>[] = Array.from({ length: SIZE }, () => new Set());
        const colRegions: Set<number>[] = Array.from({ length: SIZE }, () => new Set());
        const colorToPs = new Map<number, Region>();

        for (const ps of this.regions) {
            if (!ps.length) continue;
            const color = this.board[ps[0][0]][ps[0][1]];
            colorToPs.set(color, ps);
            for (const [y, x] of ps) {
                rowRegions[y].add(color);
                colRegions[x].add(color);
            }
        }

        const indices = Array.from({ length: SIZE }, (_, i) => i);
        for (let k = 1; k < SIZE; k++) {
            let changed = false;
            for (const rowSubset of combinations(indices, k)) {
                const regions = new Set<number>();
                let tooMany = false;
                for (const y of rowSubset) {
                    for (const c of rowRegions[y]) regions.add(c);
                    if (regions.size > k) { tooMany = true; break; }
                }
                if (tooMany || regions.size != k) continue;
                for (const color of regions) {
                    for (const [y, x] of colorToPs.get(color)!) {
                        if (!rowSubset.includes(y)) {
                            this.mask.add(hashPt(y, x));
                            changed = true;
                        }
                    }
                }
            }

            if (changed) {
                break;
            }

            for (const colSubset of combinations(indices, k)) {
                const regions = new Set<number>();
                let tooMany = false;
                for (const x of colSubset) {
                    for (const c of colRegions[x]) regions.add(c);
                    if (regions.size > k) { tooMany = true; break; }
                }
                if (tooMany || regions.size != k) continue;
                for (const color of regions) {
                    for (const [y, x] of colorToPs.get(color)!) {
                        if (!colSubset.includes(x)) this.mask.add(hashPt(y, x));
                    }
                }
            }
        }

        this.readjustRegions();
    }

    private readjustRegions() {
        for (const penset of this.regions) {
            for (let i = penset.length - 1; i >= 0; i--) {
                if (this.mask.has(hashPt(penset[i][0], penset[i][1]))) {
                    penset.splice(i, 1);
                }
            }
        }
    }

    private getRegions(): Region[] {
        const penSets: Region[] = Array.from({ length: this.size }, () => []);
        for (let row = 0; row < this.size; row++) {
            for (let col = 0; col < this.size; col++) {
                penSets[this.board[row][col]].push([row, col]);
            }
        }
        return penSets;
    }

    private getAdjacent(y: number, x: number): Point[] {
        const out: Point[] = [];
        for (let dy = -1; dy <= 1; dy++) {
            for (let dx = -1; dx <= 1; dx++) {
                if (dy == 0 && dx == 0) continue;
                const ny = y + dy, nx = x + dx;
                if (ny >= 0 && ny < this.size && nx >= 0 && nx < this.size) {
                    out.push([ny, nx]);
                }
            }
        }
        return out;
    }

    private pensetAllHoriz(penset: Region): [number, number] | null {
        const [ypos, _] = penset[0];
        const color = this.board[ypos][penset[0][1]];
        for (let i = 1; i < penset.length; i++) {
            if (penset[i][0] != ypos) return null;
        }
        return [ypos, color];
    }

    private pensetAllVert(penset: Region): [number, number] | null {
        const [_, xpos] = penset[0];
        const color = this.board[penset[0][0]][xpos];
        for (let i = 1; i < penset.length; i++) {
            if (penset[i][1] != xpos) return null;
        }
        return [xpos, color];
    }
}
